internal/era/era.go

```diff
@@ -0,0 +1,370 @@
+				}
+				next += 1
+				eras = append(eras, entry.Name())
+			}
+		}
+	}
+	return eras, nil
+}
+
+// Reader reads an Era1 archive.
+// See Builder documentation for a detailed explanation of the Era1 format.
+type Reader struct {
+	r io.ReaderAt
+	e *e2store.Reader
+
+	buf      [8]byte  // buffer reading entry offsets
+	next     uint64   // next block to read
+	length   int64    // total length of r
+	metadata metadata // start, count info
+}
+
+type ReadAtSeeker interface {
+	io.ReaderAt
+	io.Seeker
+}
+
+// NewReader returns a new Reader instance.
+func NewReader(r ReadAtSeeker) (*Reader, error) {
+	length, err := r.Seek(0, io.SeekEnd)
+	if err != nil {
+		return nil, err
+	}
+	m, err := readMetadata(r, length)
+	if err != nil {
+		return nil, err
+	}
+	return &Reader{
+		r:        r,
+		e:        e2store.NewReader(r),
+		buf:      [8]byte{},
+		next:     m.start,
+		length:   length,
+		metadata: m,
+	}, nil
+}
+
+// readOffset reads a specific block's offset from the block index. The value n
+// is the absolute block number desired.
+func (r *Reader) readOffset(n uint64) (int64, error) {
+	var (
+		firstIndex  = -8 - int64(r.metadata.count)*8      // size of count - index entries
+		indexOffset = int64(n-r.metadata.start) * 8       // desired index * size of indexes
+		offOffset   = r.length + firstIndex + indexOffset // offset of block offset
+	)
+	r.clearBuffer()
+	if _, err := r.r.ReadAt(r.buf[:], offOffset); err != nil {
+		return 0, err
+	}
+	// Since the block offset is relative from its location + size of index
+	// value (8), we need to add it to it's offset to get the block's
+	// absolute offset.
+	return offOffset + 8 + int64(binary.LittleEndian.Uint64(r.buf[:])), nil
+}
+
+// Read reads one (block, receipts) tuple from an Era1 archive.
+func (r *Reader) Read() (*types.Block, types.Receipts, error) {
+	block, receipts, err := r.ReadBlockAndReceipts(r.next)
+	if err != nil {
+		return nil, nil, err
+	}
+	r.next += 1
+	return block, receipts, nil
+}
+
+// ReadHeader reads the header number n RLP.
+func (r *Reader) ReadHeaderRLP(n uint64) ([]byte, error) {
+	if n < r.metadata.start || r.metadata.start+r.metadata.count < n {
+		return nil, fmt.Errorf("request out-of-bounds: want %d, start: %d, count: %d", n, r.metadata.start, r.metadata.count)
+	}
+	// Read the specified block's offset from the block index.
+	off, err := r.readOffset(n)
+	if err != nil {
+		return nil, fmt.Errorf("error reading block offset: %w", err)
+	}
+	// Read header.
+	var e e2store.Entry
+	if _, err := r.e.ReadAt(&e, off); err != nil {
+		return nil, err
+	}
+	if e.Type != TypeCompressedHeader {
+		return nil, fmt.Errorf("expected header entry, got %x", e.Type)
+	}
+	return io.ReadAll(snappy.NewReader(bytes.NewReader(e.Value)))
+}
+
+func skipN(r *Reader, off int64, n int) (int64, error) {
+	for i := 0; i < n; i++ {
+		_, length, err := r.e.ReadMetadataAt(off) // header
+		if err != nil {
+			return off, err
+		}
+		off += e2store.HeaderSize + int64(length)
```

Comment:
It's not great that knowledge of the format escapes to another package here. I think it'd be better to have 
a function in Reader that returns the next offset.

Reply:
This is a good point. I've fixed it by creating a method `LengthAt` which returns the overall length of the entry (including header) at the given offset.